mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp

```diff
@@ -1394,6 +1394,33 @@ class DropInnerMostUnitDimsTransferWrite
     if (dimsToDrop == 0)
       return failure();
 
+    // We need to consider 3 cases for the dim to drop:
+    //  1. if "in bounds", it can safely be assumeed that the corresponding
+    //     index is equal to 0 (safe to collapse) (*)
+    //  2. if "out of bounds" and the corresponding index is 0, it is
+    //     effectively "in bounds" (safe to collapse)
+    //  3. If "out of bounds" and the correspondong index is != 0,
+    //     be conservative and bail out (not safe to collapse)
+    // (*)  This pattern only drops unit dims, so the only possible "in bounds"
+    // index is "0". This could be added as a folder.
+    // TODO: Deal with 3. by e.g. proppaging the "out of bounds" flag to other
+    // dims.
+    bool indexOutOfBounds = true;
+    if (writeOp.getInBounds())
+      indexOutOfBounds = llvm::any_of(
+          llvm::zip(writeOp.getInBounds()->getValue().take_back(dimsToDrop),
+                    writeOp.getIndices().take_back(dimsToDrop)),
+          [](auto zipped) {
+            auto inBounds = cast<BoolAttr>(std::get<0>(zipped)).getValue();
+            auto nonZeroIdx = !isZeroIndex(std::get<1>(zipped));
+            return !inBounds && nonZeroIdx;
+          });
+    else
+      indexOutOfBounds = !llvm::all_of(
+          writeOp.getIndices().take_back(dimsToDrop), isZeroIndex);
+    if (indexOutOfBounds)
+      return failure();
```

Comment:
This looks much more complex than I'd expect. I don't see the need to check the indices at all. 

Also, dropping the dim if the index is zero and marked as out-of-bounds does not seem valid. If index zero is out-of-bounds, then we can't safely write to that unit dimension (and the write probably should be a no-op). 

Why not just:
```c++
auto inBounds = writeOp.getInBoundsValues();
auto droppedInBounds = ArrayRef<bool>(inBounds).take_back(dimsToDrop);
if (llvm::is_contained(droppedInBounds, false)) 
  return failure();
```

Reply:
(or maybe just):
```c++
if (writeOp.hasOutOfBoundsDim())
  return failure();
```
out-of-bounds dims are a edge-case (that I've not really seen used in practice, so are probably not worth worrying about).

Reply:
Side note: If index 0 and in-bounds = false for a unit-dim actually means in-bounds, that should be an xferOp fold (not something everyone that looks at in-bounds values needs to interpret).

Reply:
>  dropping the dim if the index is zero and marked as out-of-bounds does not seem valid. If index zero is out-of-bounds, then we can't safely write to that unit dimension

As you observed further down:

> If index 0 and in-bounds = false for a unit-dim actually means in-bounds = true

:) So:
1.  if the index is == 0 then `in_bounds` is effectively irrelevant (safe to collapse)
2. if index is != 0, but `in_bounds = true`, the index is effectively ==0 and  (safe to collapse)
3. if index != 0 and `in_bounds = false`, bail out.

Note that once `in_bounds` is mandatory, I will be able to simply the above as:
```cpp
      if(llvm::any_of(
          llvm::zip(writeOp.getInBounds()->getValue().take_back(dimsToDrop),
                    writeOp.getIndices().take_back(dimsToDrop)),
          [](auto zipped) {
            auto inBounds = cast<BoolAttr>(std::get<0>(zipped)).getValue();
            auto nonZeroIdx = !isZeroIndex(std::get<1>(zipped));
            return !inBounds && nonZeroIdx;
          }))
      return failure();
```

> out-of-bounds dims are an edge-case

Not sure we can say that - the default for `in_bounds` is "out of bounds".

Reply:
> It's an actual fold too (foldTransferInBoundsAttribute, so it's running pretty much all the time).

I knew about the folder, but incorrectly assumed that the "folder" wasn't guaranteed to be run before the pattern. I was wrong: https://github.com/llvm/llvm-project/blob/1ed84a862f9ce3c60251968f23a5405f06458975/mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h#L107-L108

> /// Also performs folding and simple dead-code elimination before attempting to
/// match any of the provided patterns.

So:
> Why not just:
> 
> auto inBounds = writeOp.getInBoundsValues();
> auto droppedInBounds = ArrayRef<bool>(inBounds).take_back(dimsToDrop);
> if (llvm::is_contained(droppedInBounds, false)) 
>  return failure();

Yeah, that's perfectly sufficient and covers all the cases. Thanks!

Reply:
> Not sure we can say that - the default for in_bounds is "out of bounds".
I think this is definitely safer.